src/cljx/schema/core.cljx

```diff
@@ -515,6 +487,83 @@
      [(if (= pred :else) (constantly true) pred) schema])
    (if (odd? (count preds-and-schemas)) (last preds-and-schemas))))
 
+(defprotocol HasPrecondition
+  (precondition [this]
+    "Return the Precondition for this schema.
+     A Precondition is a function of a value that returns a
+     ValidationError or nil (see spec.core)"))
+
+(extend-protocol HasPrecondition
+  schema.spec.leaf.LeafSpec
+  (precondition [this]
+    (:pre this))
+
+  schema.spec.variant.VariantSpec
+  (precondition [this]
+    (let [{:keys [pre options]} this]
+      (some-fn
+       pre
+       (spec/precondition
+        Any
+        (complement
+         (apply every-pred
+                (for [{:keys [guard schema]} options]
+                  (if guard
+                    (some-fn (spec/precondition schema guard #(list 'guard %))
+                             (precondition (spec schema)))
+                    (precondition (spec schema))))))
+        #(list 'some-variant-precondition %)))))
+
+  schema.spec.collection.CollectionSpec
+  (precondition [this]
+    (:pre this)))
+
+(clojure.core/defn cond-pre
+  "A replacement for `either` that constructs a conditional schema
+   based on the schema spec preconditions of the component schemas.
+
+   EXPERIMENTAL"
+  [& schemas]
+  (apply conditional
+         (for [s schemas
+               t [(comp not (precondition (spec s))) s]]
+           t)))
```

Comment:
:sheep: might be clearer to do `(apply concat (for [s schemas] [(comp ...) s]))`